dev-docs/batch-system.rst

```diff
@@ -0,0 +1,784 @@
+MySQL Database
+~~~~~~~~~~~~~~
+
+All Batch and Auth state is stored in a cloud-provider managed MySQL
+database. We use SSL certificates to secure communication between
+Kubernetes services and the database. Worker VMs cannot talk directly
+to the database.
+
+
+~~~~~~~~~~~~~
+Cloud Storage
+~~~~~~~~~~~~~
+
+Users store the data they want to compute on in Cloud Storage (Google
+Cloud Storage or Azure Blob Storage). All Batch created files such as
+user job specs, job log files, job status files, and job resource
+usage monitoring files are stored in cloud storage.
+
+
+~~~~~~~~~~~~~~~~~~
+Container Registry
+~~~~~~~~~~~~~~~~~~
+
+Container images used to execute user jobs as well as the images used
+in our Kubernetes services are stored in a cloud provider managed
+Container Registry (Google Artifact Registry and Azure Container
+Registry).
+
+
+~~~~~~~~~
+Terraform
+~~~~~~~~~
+
+TBD.
+
+
+~~~~~~~~~~~~~
+Bootstrapping
+~~~~~~~~~~~~~
+
+TBD.
+
+
+===================
+Application Details
+===================
+
+~~~~~~~~~~~~~~~
+Batch Lifecycle
+~~~~~~~~~~~~~~~
+
+1. A user submits a request to the Batch front end service to create a
+   batch along with job specifications.
+2. The Batch front end service records the batch and job information
+   into a MySQL database and writes the job specifications to cloud
+   storage.
+3. The Batch driver notices that there is work available either
+   through a push request from the Batch front end or by polling the
+   state in the MySQL database and spins up worker VMs.
+4. The worker VMs startup and notify the Batch driver they are active
+   and have resources to run jobs.
+5. The Batch driver schedules jobs to run on the active workers.
+6. The worker VM downloads the job specification from cloud storage,
+   downloads any input files the job needs from cloud storage, creates
+   a container for the job to execute in, executes the code inside the
+   container, uploads any logs and output files that have been
+   generated, and then notifies the Batch driver that the job has
+   completed.
+7. Once all jobs have completed, the batch is set to complete in the
+   database. Any callbacks that have been specified on batch
+   completion are called.
+8. Meanwhile, the user can find the status of their batch through the
+   UI or using a Python client library to get the batch status, cancel
+   the batch, list the jobs in the batch and their statuses, and wait
+   for the batch or an individual job to complete. The implementation
+   of the wait operation is by continuously polling the Batch Front
+   End until the batch state is "complete".
+
+
+~~~~~~~~~~
+Data Model
+~~~~~~~~~~
+
+The core concepts in the Batch data model are billing projects,
+batches, jobs, updates, attempts, and resources.
+
+A **billing project** is a mechanism for imposing cost control and
+enabling the ability to share information about batches and jobs
+across users. Each billing project has a list of authorized users and
```

Comment:
```suggestion
A **billing project** is a mechanism for cost accounting, cost control, and
enabling the ability to share information about batches and jobs
across users. Each billing project has a list of authorized users and
```

Billing projects are like credit cards or some other account to which we can assign a bill. I agree they also serve as a locus of cost control.